### ISSUE TITLE
Out of memory exception when using OsmSimpleCompleteStreamSource

### ISSUE THREAD

**Issue by HarelM** [OsmSharp/core#7]
Hi,
I'm trying to use the complete stream source on a relatively large data (Entire Isreal)
http://download.geofabrik.de/asia/israel-and-palestine-latest.osm.pbf
The following is the exception: 
``` 
at OsmSharp.Osm.PBF.Streams.PBFOsmStreamSource.MoveNext(Boolean ignoreNodes, Boolean ignoreWays, Boolean ignoreRelations)
 at OsmSharp.Osm.Streams.Complete.OsmSimpleCompleteStreamSource.MoveNext()
 at IsraelHiking.API.Services.OsmDataService.<UpdateElasticSearchFromFile>d__11.MoveNext() in D:\Github\IsraelHikingMap\Site\IsraelHiking.API\Services\OsmDataService.cs:line 82
```
I'm using a machine with 16Gb RAM.
I tried to use it without the complete stream and I think it didn't had this exception, but then I needed to cache the data into dictionaries and I think the issue was there when the dictionaries got too big.
the code I'm writing is something like so:
```
            var source = new PBFOsmStreamSource(stream);
            await _elasticSearchGateway.DeleteAll();
            var completeSource = new OsmSimpleCompleteStreamSource(source);
            var converter = new OsmGeoJsonConverter();
            var list = new List<Feature>(PAGE_SIZE);
            int page = 0;
            foreach (var completeOsmGeo in completeSource)
            {
                var geoJson = converter.ToGeoJson(completeOsmGeo);
                if (geoJson?.Properties == null || !geoJson.Properties.Keys.Any(k => k.Contains("name")))
                {
                    continue;
                }
                list.Add(geoJson);
                if (list.Count == PAGE_SIZE)
                {
                    page++;
                    _logger.Info($"Indexing {PAGE_SIZE * page} records");
                    _elasticSearchGateway.UpdateData(list).Wait();
                    list.Clear();
                }
            }
```
basically what it does is convert the osm to geojson and index in bulks in elastic search.
Any advise?

**Comment by xivk** [comment]
Much too late but i'm going to anwser anyway. One piece of advice, filter the objects _before_ the reach the complete stream. Meaning add a filter to keep only objects with a 'name' tag. If this still is too much data there are 3 options:

- Process the data in different stages. For example process ways/relations with ID's mod 4 and thus in 4 stages. This way the process needs less memory and just loops over the source stream 4 times. Speed-memory tradeoff.
- Implement a more efficient ISnapshotDb. You can use this as an argument in the constructor of OsmSimpleCompleteStreamSource. You can even use a cache based on SQLite or use another database or disk to cache objects. There is no default implement in OsmSharp for this at the moment but there is an [sqlite example](https://github.com/OsmSharp/sqlite).
- Don't use the complete stream and implement your own multi-stage processor. Like in Itinero for routing data: https://github.com/itinero/routing/blob/develop/Itinero.IO.Osm/Streams/RouterDbStreamTarget.cs

**Comment by HarelM** [comment]
In this case the problem was that the project file had a flag called "prefer 32 bit process" checked (which is checked by default in visual studio).
When I unchecked this flag the issue was fixed.

**Comment by xivk** [comment]
ok :+1:
